feat: implement write syscall in engine (#20)

mstarzinger opened the pull request.

ChristianMoesl commented on src/engine/mod.rs:
```diff
@@ -564,6 +564,36 @@ where
+        if !self.is_in_vaddr_range(buffer) || !self.is_in_vaddr_range(buffer + size) {
+            return not_supported("write syscall failed to access buffer");
+        }
+
```
I think we should probably try to actually access the buffer, just to see if there are "Uninitialized" values in there. Usually that's defined as undefined behaviour and therefore I think we should interpret it is a bug. (https://github.com/cksystemsgroup/monster/blob/main/src/engine/bug.rs#L14)

mstarzinger replied:
Good point. I'll add a loop that scans for uninitialized words in the given buffer.

One quirk I see here is that there will potentially be a strange interaction with the way `read` handles partially initialized words. IIUC, then `read` will mark the entire word as uninitialized if at least one byte remains uninitialized. So the following C* snippet will always fail I think:

```
uint64_t* x;
x = malloc(8);
read(0, x, 7);
write(1, x, 7);
```

Do you have thoughts on that? I am unclear what the semantics of partially initialized words should be here.

ChristianMoesl replied:
I decided to define partially initialised words as uninitialised, because it makes everything simpler as long as we do not support other datatypes. But Christoph wants us to extend the support to full blown RISC-V, where this problem has to be resolved by then anyway.

mstarzinger replied:
Done. I have implemented the scan for (partially) uninitialized words, which will be reported as `Bug::AccessToUnitializedMemory` now.

ChristianMoesl replied:
perfect

ChristianMoesl commented on src/engine/mod.rs:
```diff
@@ -564,6 +564,36 @@ where
+        trace!("write: fd={} buffer={:#x} size={}", 1, buffer, size,);
+
+        if !self.is_in_vaddr_range(buffer) || !self.is_in_vaddr_range(buffer + size) {
+            return not_supported("write syscall failed to access buffer");
```
This is an actual bug in the input program, rather than an unsupported feature in the execution engine. Have a look at this: https://github.com/cksystemsgroup/monster/blob/main/src/engine/bug.rs#L25

mstarzinger replied:
Yeah, I was stumped by the same thing when reading the existing implementation of `read`. But looking at the [error section](https://man7.org/linux/man-pages/man2/write.2.html#ERRORS) of the man page I saw `EFAULT` as one of the handled error cases. So I think writing from an out-of-bounds buffer will just make `write` return -1 and not crash. Should this still be considered a bug in the input program by the engine? WDYT?

However we decide here I feel that we should keep the implementations of `read` and `write` in sync here and either treat this case as a bug in both or neither.

I also double-checked my understanding with the following small C (not C*) snippet, wich indeed prints `res:-1 (errno:14, 'Bad address')` when run:

```
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

int main() {
  void* x = (void*) 23;
  int res = write(0, x, 10);
  printf("res:%d (errno:%d, '%s')\n", res, errno, strerror(errno));
}
```

ChristianMoesl replied:
Hmm interesting, I did not know that. 

So we have two possibly scenarios. Either `-1` is returned or the engine has found a bug at that point. From a programmers perspective I think I would classify it as a bug. Keep in mind, that undefined behaviour is always a subset of our bug definition. It really comes down to, what we decide to be a bug. `exit(1)` is also a bug in our current view.

mstarzinger replied:
Done. This is now treated as a bug for both `read` and `write` and will be reported as `Bug::AccessToOutOfRangeAddress` in both cases.

ChristianMoesl replied:
nice

ChristianMoesl commented on src/engine/mod.rs:
```diff
@@ -564,6 +564,36 @@ where
+            return not_supported("write syscall failed to access buffer");
+        }
+
+        self.regs[Register::A0 as usize] = Value::Concrete(size);
```
This is a good default implementation I think. The only downside is: by returning the written size, you basically say, that the write syscall is always successful (if inputs are valid). What do you think about returning a symbolic value? In that way we would also explore paths in the user program, where the write syscall "failed". The symbolic value in that case has 3 valid concrete values {0, -1, size}. (https://man7.org/linux/man-pages/man2/write.2.html)

mstarzinger replied:
Good idea. I am fine extending this to other return values. In fact I think the result can actually take any value in the range [0; size] since partial writes are also possible.

Would you be OK with tackling this extension as part of a separate PR since this PR is getting big as it is? I am a huge fan of smaller incremental changes.

ChristianMoesl replied:
yeah, sounds good. Small changes are definitely preferable!

ChristianMoesl reviewed (changes requested):
@mstarzinger I have one little nit-picky change request for you. Can you look at your commit "types" (https://www.conventionalcommits.org/en/v1.0.0/) and use a more appropriate one, where possible. It's important for us, because we use a release mechanism, where we generate a changelog/increment version automatically based on commit messages. `feat` increases the minor version and `fix` the patch version. All the other types do not increase the version at all. So these types have to be chosen based on a monster library/binary user's perspective.

Btw. I messed up your PR with my latest change. As soon as you rebase your branch and update your messages, we can merge it.

mstarzinger commented:
@ChristianMoesl Good to know about the impact of the various commit "types" on the version number, I wasn't aware of that. Thanks for letting me know. As discussed offline I just squashed the entire PR into one single commit and rebased it. Should be ready to go now.

ChristianMoesl commented:
@mstarzinger looks good, thanks!
